src/hotspot/share/opto/loopopts.cpp

```diff
@@ -1503,19 +1503,32 @@ void PhaseIdealLoop::try_sink_out_of_loop(Node* n) {
           assert(!n_loop->is_member(get_loop(x_ctrl)), "should have moved out of loop");
           register_new_node(x, x_ctrl);
 
-          if (x->in(0) == NULL && !x->is_DecodeNarrowPtr()) {
+          // Chain of AddP: (AddP base (AddP base )) must keep the same base after sinking. We don't add a CastPP here
+          // when the first one is sunk so if the second one is not, their bases remain the same. A CastPP of the base
+          // is only added once both AddP nodes are sunk (see special code for AddP after the cast is created).
+          if (x->in(0) == NULL && !x->is_DecodeNarrowPtr() &&
+              !(x->is_AddP() && x->in(AddPNode::Address)->is_AddP() && x->in(AddPNode::Address)->in(AddPNode::Base) == x->in(AddPNode::Base))) {
             assert(!x->is_Load(), "load should be pinned");
             // Use a cast node to pin clone out of loop
             Node* cast = NULL;
             for (uint k = 0; k < x->req(); k++) {
               Node* in = x->in(k);
               if (in != NULL && n_loop->is_member(get_loop(get_ctrl(in)))) {
                 const Type* in_t = _igvn.type(in);
-                cast = ConstraintCastNode::make_cast_for_type(x_ctrl, in, in_t);
+                cast = ConstraintCastNode::make_cast_for_type(x_ctrl, in, in_t, ConstraintCastNode::UnconditionalDependency);
               }
               if (cast != NULL) {
                 register_new_node(cast, x_ctrl);
                 x->replace_edge(in, cast);
+                if (x->is_AddP() && k == AddPNode::Base) {
```

Comment:
Does the code assume that "inner" node is processed first?
Otherwise, the "outer" node could be already processed and bases don't match anymore.

I'm asking because `PhaseIdealLoop::split_if_with_blocks()` (which calls `split_if_with_blocks_post()` and `try_sink_out_of_loop()` does process users first.

Reply:
The expectation is that uses are processed first.  The:
!(x->is_AddP() && x->in(AddPNode::Address)->is_AddP() && x->in(AddPNode::Address)->in(AddPNode::Base) == x->in(AddPNode::Base))
check guarantees the "outer" node doesn't change the base when it's processed. When the "inner" node is processed, bases for the inner and outer node are updated.

Reply:
A comment briefly explaining what's happening here may help as well.